### DISCUSSION TITLE
Queues, releasing and rate limits and suppression lists counters

### DISCUSSION THREAD

**Discussion by dragoangel** [postalserver/postal#1516] [Development discussions]
Want to discuss questions about future of queues. What about automation of releasing mail which was rate limited? There should be then a option to choose which way to work, auto or manual. Not think that people want to go end do release manually all the time. Point of limitations are to not extended sone amount of email per x time, but not block them totally. Also as postal is designed in first for mass mail it will be really cool future to handle how much can be send to one destination by two independent parameters, first is TO: and second is: mx domain with option for MX ends with to handle stuff like Outlook. If TO: is set and it has policy for mx as well, then TO: take advantage and do policy rate limits. What you think? Limit could be global per company and per server, which one reached first - put messages in help for time limit not reset. Also I saw and think that limits good to be in type of: sec,min,hours,days,month and there could be multiple limits at once. E.g.:
1. not more then 10 per 10 seconds
2. not more then 100 per 5 minutes
3. not more then 300 per hour
4. not more then 10k per month

Minimum counter could be equal to 10s and before process email there will be check on:
1.  any of limits on server and org level is reaching?
2. if they not reached - maybe email in suppression list?

Any faced limit - held email. For each limit there will be timer to recheck if queue is free and as example we got 10 free emails in queue - oldest 10 emls from queue will be restored to delivery. Also if email delivery fails with perm error counter for this email takes back to queue and we can send another email.

Also as I started speak about hard fails - in both my setups there strange behavior how suppression list working - each email should be triggered 18 times, not all in summary. I understand why it so: as 18 repeats of soft fail  actually not a much, I saw yahoo who delay email via soft reject for a period of 6 hours and in end yahoo accepted that email. But for hard bounce same value of repeats looking bigger than it should to be. Maybe better split soft from hard to 2 different handling systems and counters with own limits?

**Comment by dragoangel** [comment]
Hi @adamcooke any update on it? As now with current handling rate limit its deniel of service function, not more :(

**Reply by adamcooke** [reply]
No updates I'm afraid. This isn't something we need on our systems so it's not something I have time to work on. I'm happy to look at any pull requests though.

**Reply by dragoangel** [reply]
Hi, you mean you never put rate limits on production system? There was pr to auto release messages from queue if that not development server and it was closed, so maybe it can be merged now? Many people are asking that. I was running 2 systems with that fix without any issues. Of course that not best implementation (no settings, just basic release logic), but it was at least working. Better something working without customization then not working at all, are you agree?:)
